File: String/isAnagram.py
'''
An anagram is a word, phrase, or name formed by rearranging the letters of another, such as cinema, formed from iceman.
Write a program to detect if a string is an anagram of another input string.

This solution is implementation of naive approach
O(N^2)
'''

def isAnagram(str1, str2):
    marked = []
    for s1 in str1:
        matchFound = False
        for i, s2 in enumerate(str2):
            # check the unmarked letter
            if i not in marked:
                # if found match, mark it and move onto next str1 letter
                if s1 == s2:
                    marked.append(i)
                    matchFound = True
                    break
        # after str2 loop, no match found
        if not matchFound:
            return False
    # after str1 loop, all str2 has been marked
    return len(marked) == len(str2)
        

def main():
    str1 = "abcdefg"
    str2 = "gedfacb"

    if isAnagram(str1, str2):
        print("%s is an anagram of %s" %(str1.upper(), str2.upper()))
    else:
        print("%s is NOT an anagram of %s" %(str1.upper(), str2.upper()))


if __name__ == "__main__":
    main()

File: String/isAnagram.cpp
/** 
 *  An anagram is a word, phrase, or name formed by rearranging the letters of another, such as cinema, formed from iceman.
 *  Write a program to detect if a string is an anagram of another input string.
 * 
 * This solution is implementation of improved approach using dictionary (AKA map in C++)
 */

#include <iostream>
#include <unordered_map>
#include <string>
using namespace std;

bool isAnagram(string str1, string str2);

int main(int argc, char const *argv[])
{
    string str1 = "abcdefg";
    string str2 = "fcegabd";

    if (isAnagram(str1, str2)) {
        cout << str1 << " is an anagram of " << str2 << endl;
    } else {
        cout << str1 << " is NOT an anagram of " << str2 << endl;
    }

    return 0;
}

bool isAnagram(string str1, string str2) {
    unordered_map<char, int> freq;
    for (char c: str1) {
        if (freq.find(c) == freq.end()) {
            freq.insert(make_pair(c, 1));
        } else {
            freq[c] += 1;
        }
    }
    for (char c: str2) {
        // if c in str2 does not exist in frequency map OR if count is already 0
        if (freq.find(c) == freq.end() || freq[c] <= 0) {
            return false;
        } else {
            freq[c] -= 1;
        }
    }
    // check to make sure all count has been decremented to 0
    for (auto it: freq) {
        if (it.second > 0)
            return false;
    }
    return true;
}
